File: sdk-ios/Sources/WailoSDK/WailoCrypto.swift
import CommonCrypto
import CryptoKit
import Foundation

/// The primitives behind the identity-first WiFi handshake (ADR-0060), and the one place the wire's
/// byte layouts are decided. `sdk-android` and `engine` reimplement all of this against
/// `java.security`, so every
/// constant, label and concatenation order here is a contract: change one and paired devices stop
/// authenticating with no useful error. The cross-platform vectors in the test suites exist to catch
/// exactly that drift.
///
/// The two sides prove themselves differently because their exposure differs. Studio signs with an
/// ECDSA P-256 key, so nothing a device stores — nor anything that leaks out of Studio's own key
/// store — can forge that identity. A device answers with an HMAC over a paired secret, which is
/// enough when it only has to convince the one Studio it paired with.
///
/// P-256 rather than Ed25519 purely for reach: `java.security` only exposes Ed25519 at Android API 33
/// and the SDK's floor is 24, whereas `SHA256withECDSA` has been in the platform since API 11 and
/// `P256.Signing` since iOS 13. ECDSA's real hazard is a reused or biased signing nonce leaking the
/// private key, and it cannot reach the device: only Studio signs, and verification uses no nonce.
///
/// Keys cross the wire as X9.63 uncompressed points and signatures as raw `r || s`, because those are
/// the representations CryptoKit exposes at iOS 13 — DER needs 14. Studio converts to and from the
/// JVM's native SPKI/DER on its own side.
enum WailoCrypto {

    /// Both handshake nonces, and every derived key.
    static let nonceLength = 32
    static let keyLength = 32

    // MARK: - identity

    /// Studio's identity: the first 16 bytes of its public key's SHA-256, in lowercase hex.
    ///
    /// Deriving it from the key rather than picking a random string is what makes the Bonjour TXT
    /// record self-authenticating. An impostor can advertise someone else's `sid`, but it would need a
    /// key that hashes to it in order to sign StudioHello — so the device dials it, fails to verify,
    /// and hangs up, instead of racing the real Studio for the device's attention.
    ///
    /// The key is hashed in its X9.63 form (the 65-byte uncompressed point), not X.509 SPKI DER, so
    /// both platforms hash the same bytes. Studio converts on its side; CryptoKit only offers DER
    /// above iOS 14 and this package targets 13.
    static func studioId(publicKey: Data) -> String {
        Data(SHA256.hash(data: publicKey).prefix(16)).hexadecimal
    }

    // MARK: - key agreement

    /// A fresh P-256 agreement pair, one per connection and discarded with it.
    static func generateEphemeral() -> P256.KeyAgreement.PrivateKey {
        P256.KeyAgreement.PrivateKey()
    }

    /// The agreed secret: the X coordinate of the shared point, which is what CryptoKit and the JVM's
    /// `ECDH` both return. Nil for a peer key that is malformed or off the curve — CryptoKit checks
    /// that on the way in, so an invalid-curve probe fails here rather than agreeing to something.
    static func agree(_ privateKey: P256.KeyAgreement.PrivateKey, peer: Data) -> Data? {
        guard let peerKey = try? P256.KeyAgreement.PublicKey(x963Representation: peer),
              let secret = try? privateKey.sharedSecretFromKeyAgreement(with: peerKey) else { return nil }
        return secret.withUnsafeBytes { Data($0) }
    }

    // MARK: - key schedule

    static func deviceKeyV3(pairingSecret: Data, studioId: String, deviceAlias: String) -> SymmetricKey {
        hkdf(
            keyMaterial: pairingSecret,
            salt: Data(studioId.utf8),
            info: label("wailo/device-key/v3", Data(deviceAlias.utf8))
        )
    }

    static func tofuDeviceKeyV3(shared: Data, studioId: String, deviceAlias: String) -> SymmetricKey {
        hkdf(
            keyMaterial: shared,
            salt: Data(studioId.utf8),
            info: label("wailo/tofu-device-key/v3", Data(deviceAlias.utf8))
        )
    }

    static func authKeyV3(shared: Data, deviceKey: SymmetricKey) -> SymmetricKey {
        hkdf(
            keyMaterial: shared + deviceKey.bytes,
            salt: Data(),
            info: Data("wailo/auth/v3".utf8)
        )
    }

    static func sessionKeyV3(
        shared: Data,
        deviceKey: SymmetricKey,
        nonceD: Data,
        nonceS: Data
    ) -> SymmetricKey {
        hkdf(
            keyMaterial: shared + deviceKey.bytes,
            salt: nonceD + nonceS,
            info: Data("wailo/session/v3".utf8)
        )
    }

    // MARK: - proofs

    /// Stretches a typed pairing code into the same 32 bytes a QR would have carried directly.
    ///
    /// A code short enough to type holds ~50 bits, and everything downstream descends from it, so
    /// anyone who records one pairing handshake could otherwise brute-force it offline and decrypt
    /// every session that device ever has. The iteration count is the mitigation: it makes each guess
    /// cost real work. A PAKE would let the code shrink further and is the proper fix if the typing
    /// ever grates.
    static func stretch(code: String, studioId: String) -> Data {
        pbkdf2(password: Data(code.utf8), salt: Data(studioId.utf8), rounds: 200_000)
    }

    /// Constant-time, because these compare secrets and `==` on `Data` returns early.
    static func constantTimeEquals(_ lhs: Data, _ rhs: Data) -> Bool {
        guard lhs.count == rhs.count else { return false }
        var difference: UInt8 = 0
        for (left, right) in zip(lhs, rhs) { difference |= left ^ right }
        return difference == 0
    }

    static func deviceProofV3(
        authKey: SymmetricKey,
        studioId: String,
        nonceD: Data,
        nonceS: Data,
        ephemeralD: Data,
        ephemeralS: Data,
        pairingRequired: Bool,
        deviceAlias: String,
        mode: Int32,
        sessionCounter: UInt64
    ) -> Data {
        let transcript = authTranscriptV3(
            role: "wailo/device-proof/v3",
            studioId: studioId,
            nonceD: nonceD,
            nonceS: nonceS,
            ephemeralD: ephemeralD,
            ephemeralS: ephemeralS,
            pairingRequired: pairingRequired,
            deviceAlias: deviceAlias,
            mode: mode,
            sessionCounter: sessionCounter
        )
        return Data(HMAC<SHA256>.authenticationCode(for: transcript, using: authKey))
    }

    static func studioProofV3(
        authKey: SymmetricKey,
        studioId: String,
        nonceD: Data,
        nonceS: Data,
        ephemeralD: Data,
        ephemeralS: Data,
        pairingRequired: Bool,
        deviceAlias: String,
        mode: Int32,
        sessionCounter: UInt64,
        resultCode: Int32
    ) -> Data {
        let transcript = resultTranscriptV3(
            role: "wailo/studio-proof/v3",
            studioId: studioId,
            nonceD: nonceD,
            nonceS: nonceS,
            ephemeralD: ephemeralD,
            ephemeralS: ephemeralS,
            pairingRequired: pairingRequired,
            deviceAlias: deviceAlias,
            mode: mode,
            sessionCounter: sessionCounter,
            resultCode: resultCode
        )
        return Data(HMAC<SHA256>.authenticationCode(for: transcript, using: authKey))
    }

    static func isValidStudioHelloV3(
        _ signature: Data,
        publicKey: Data,
        studioId: String,
        nonceD: Data,
        nonceS: Data,
        ephemeralD: Data,
        ephemeralS: Data,
        pairingRequired: Bool
    ) -> Bool {
        verifyV3(
            signature: signature,
            publicKey: publicKey,
            transcript: helloTranscriptV3(
                role: "wailo/studio-hello/v3",
                studioId: studioId,
                nonceD: nonceD,
                nonceS: nonceS,
                ephemeralD: ephemeralD,
                ephemeralS: ephemeralS,
                pairingRequired: pairingRequired
            )
        )
    }

    static func isValidResultSignatureV3(
        _ signature: Data,
        publicKey: Data,
        studioId: String,
        nonceD: Data,
        nonceS: Data,
        ephemeralD: Data,
        ephemeralS: Data,
        pairingRequired: Bool,
        deviceAlias: String,
        mode: Int32,
        sessionCounter: UInt64,
        resultCode: Int32
    ) -> Bool {
        verifyV3(
            signature: signature,
            publicKey: publicKey,
            transcript: resultTranscriptV3(
                role: "wailo/result-signature/v3",
                studioId: studioId,
                nonceD: nonceD,
                nonceS: nonceS,
                ephemeralD: ephemeralD,
                ephemeralS: ephemeralS,
                pairingRequired: pairingRequired,
                deviceAlias: deviceAlias,
                mode: mode,
                sessionCounter: sessionCounter,
                resultCode: resultCode
            )
        )
    }

    static func randomNonce() -> Data {
        var bytes = Data(count: nonceLength)
        bytes.withUnsafeMutableBytes { buffer in
            guard let base = buffer.baseAddress else { return }
            _ = SecRandomCopyBytes(kSecRandomDefault, nonceLength, base)
        }
        return bytes
    }

    // MARK: - internals

    /// A `0x00`-separated label, so a payload can never be read two ways. Every field either side of a
    /// separator is fixed-length or terminated by one, which is what keeps the transcript unambiguous
    /// across two independent implementations.
    private static func label(_ text: String, _ suffix: Data) -> Data {
        Data(text.utf8) + Data([0x00]) + suffix + Data([0x00])
    }

    private static func helloTranscriptV3(
        role: String,
        studioId: String,
        nonceD: Data,
        nonceS: Data,
        ephemeralD: Data,
        ephemeralS: Data,
        pairingRequired: Bool
    ) -> Data {
        label(role, Data(studioId.utf8)) +
            nonceD + nonceS + ephemeralD + ephemeralS + Data([pairingRequired ? 1 : 0])
    }

    private static func authTranscriptV3(
        role: String,
        studioId: String,
        nonceD: Data,
        nonceS: Data,
        ephemeralD: Data,
        ephemeralS: Data,
        pairingRequired: Bool,
        deviceAlias: String,
        mode: Int32,
        sessionCounter: UInt64
    ) -> Data {
        helloTranscriptV3(
            role: role,
            studioId: studioId,
            nonceD: nonceD,
            nonceS: nonceS,
            ephemeralD: ephemeralD,
            ephemeralS: ephemeralS,
            pairingRequired: pairingRequired
        ) + label("wailo/device-alias/v3", Data(deviceAlias.utf8)) +
            fixed(mode) + fixed(sessionCounter)
    }

    private static func resultTranscriptV3(
        role: String,
        studioId: String,
        nonceD: Data,
        nonceS: Data,
        ephemeralD: Data,
        ephemeralS: Data,
        pairingRequired: Bool,
        deviceAlias: String,
        mode: Int32,
        sessionCounter: UInt64,
        resultCode: Int32
    ) -> Data {
        authTranscriptV3(
            role: role,
            studioId: studioId,
            nonceD: nonceD,
            nonceS: nonceS,
            ephemeralD: ephemeralD,
            ephemeralS: ephemeralS,
            pairingRequired: pairingRequired,
            deviceAlias: deviceAlias,
            mode: mode,
            sessionCounter: sessionCounter
        ) + fixed(resultCode)
    }

    private static func fixed(_ value: Int32) -> Data {
        var bigEndian = value.bigEndian
        return withUnsafeBytes(of: &bigEndian) { Data($0) }
    }

    private static func fixed(_ value: UInt64) -> Data {
        var bigEndian = value.bigEndian
        return withUnsafeBytes(of: &bigEndian) { Data($0) }
    }

    private static func verifyV3(signature: Data, publicKey: Data, transcript: Data) -> Bool {
        guard let key = try? P256.Signing.PublicKey(x963Representation: publicKey),
              let signature = try? P256.Signing.ECDSASignature(rawRepresentation: signature) else {
            return false
        }
        return key.isValidSignature(signature, for: transcript)
    }

    /// RFC 5869 HKDF-SHA256, hand-composed because CryptoKit's own `HKDF` is iOS 14 and this package
    /// targets iOS 13. Only the composition is ours — extract and expand are both HMAC, which is the
    /// platform's. Output is one SHA-256 block, so expand needs a single iteration.
    private static func hkdf(keyMaterial: Data, salt: Data, info: Data) -> SymmetricKey {
        let prk = HMAC<SHA256>.authenticationCode(for: keyMaterial, using: SymmetricKey(data: salt))
        let block = HMAC<SHA256>.authenticationCode(
            for: info + Data([0x01]),
            using: SymmetricKey(data: Data(prk))
        )
        return SymmetricKey(data: Data(block))
    }

    /// CommonCrypto rather than CryptoKit, which has no PBKDF2 at any deployment target.
    private static func pbkdf2(password: Data, salt: Data, rounds: UInt32) -> Data {
        var derived = Data(count: keyLength)
        let status: Int32 = derived.withUnsafeMutableBytes { output in
            password.withUnsafeBytes { password in
                salt.withUnsafeBytes { salt in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        password.baseAddress?.assumingMemoryBound(to: CChar.self), password.count,
                        salt.baseAddress?.assumingMemoryBound(to: UInt8.self), salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256), rounds,
                        output.baseAddress?.assumingMemoryBound(to: UInt8.self), keyLength
                    )
                }
            }
        }
        // Only a nonsensical parameter can fail here, and every one of them is a constant above.
        precondition(status == kCCSuccess, "PBKDF2 derivation failed")
        return derived
    }
}

extension SymmetricKey {
    var bytes: Data { withUnsafeBytes { Data($0) } }
}

extension Data {
    var hexadecimal: String { map { String(format: "%02x", $0) }.joined() }

    init?(hexadecimal text: String) {
        guard text.count.isMultiple(of: 2) else { return nil }
        var bytes = Data(capacity: text.count / 2)
        var index = text.startIndex
        while index < text.endIndex {
            let next = text.index(index, offsetBy: 2)
            guard let byte = UInt8(text[index..<next], radix: 16) else { return nil }
            bytes.append(byte)
            index = next
        }
        self = bytes
    }
}

File: sdk-android/src/main/kotlin/com/venbiasa/wailo/sdk/android/WailoCrypto.kt
package com.venbiasa.wailo.sdk.android

import java.math.BigInteger
import java.nio.ByteBuffer
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECFieldFp
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * The device's half of the identity-first WiFi handshake primitives (ADR-0060), the third independent
 * implementation of one byte-for-byte contract: `sdk-ios/Sources/WailoSDK/WailoCrypto.swift` and
 * `engine/pairing/WailoCrypto.kt`. Nothing can be shared with either — `sdk-ios` is Swift, and the
 * studio build is a separate Gradle build joined only by the `protocol` artifact (ADR-0015) — so every
 * label, constant and concatenation order here is a contract, and `WailoCryptoTest` asserts the same
 * vectors all three suites do. Drift shows up in the field as a device that pairs and then silently
 * fails to authenticate, with nothing in any log to say why.
 *
 * Only the device's half is here. Studio signs and the device verifies, so there is no `sign` and no
 * DER-encoder — the conversion runs one way, raw `r || s` to DER, purely to feed `SHA256withECDSA`.
 */
internal object WailoCrypto {

    const val NONCE_LENGTH = 32
    const val KEY_LENGTH = 32

    private const val CURVE = "secp256r1"
    private const val SIGNATURE_ALGORITHM = "SHA256withECDSA"
    private const val HMAC = "HmacSHA256"
    private const val COORDINATE_LENGTH = 32
    private const val PBKDF2_ROUNDS = 200_000

    private val random = SecureRandom()
    private val TWO = BigInteger.valueOf(2)
    private val THREE = BigInteger.valueOf(3)

    fun randomNonce(): ByteArray = ByteArray(NONCE_LENGTH).also(random::nextBytes)

    fun randomSecret(): ByteArray = ByteArray(KEY_LENGTH).also(random::nextBytes)

    // MARK: identity

    /** The first 16 bytes of the X9.63 public key's SHA-256, lowercase hex. */
    fun studioId(publicKey: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(publicKey).copyOf(16).toHex()

    // MARK: key agreement

    /** A fresh P-256 agreement pair, one per connection and discarded with it. */
    fun generateEphemeral(): KeyPair =
        KeyPairGenerator.getInstance("EC")
            .apply { initialize(ECGenParameterSpec(CURVE), random) }
            .generateKeyPair()

    /**
     * The agreed secret: the X coordinate of the shared point, which is what the JVM's `ECDH` and
     * CryptoKit both return. Null for a peer key that is malformed or off the curve.
     */
    fun agree(privateKey: PrivateKey, peer: ByteArray): ByteArray? = runCatching {
        KeyAgreement.getInstance("ECDH").apply {
            init(privateKey)
            doPhase(decodePublicKey(peer), true)
        }.generateSecret()
    }.getOrNull()

    // MARK: key schedule

    fun deviceKeyV3(pairingSecret: ByteArray, studioId: String, deviceAlias: String): ByteArray =
        hkdf(pairingSecret, studioId.toByteArray(), label("wailo/device-key/v3", deviceAlias.toByteArray()))

    fun tofuDeviceKeyV3(shared: ByteArray, studioId: String, deviceAlias: String): ByteArray =
        hkdf(shared, studioId.toByteArray(), label("wailo/tofu-device-key/v3", deviceAlias.toByteArray()))

    fun authKeyV3(shared: ByteArray, deviceKey: ByteArray): ByteArray =
        hkdf(shared + deviceKey, ByteArray(0), "wailo/auth/v3".toByteArray())

    fun sessionKeyV3(
        shared: ByteArray,
        deviceKey: ByteArray,
        nonceD: ByteArray,
        nonceS: ByteArray,
    ): ByteArray = hkdf(shared + deviceKey, nonceD + nonceS, "wailo/session/v3".toByteArray())

    /**
     * Stretches a typed pairing code into the same 32 bytes a scanned QR carries directly. The
     * iteration count is what makes a code short enough to type survive an offline guess.
     */
    fun stretch(code: String, studioId: String): ByteArray =
        pbkdf2(code.toByteArray(), studioId.toByteArray(), PBKDF2_ROUNDS)

    // MARK: proofs

    fun deviceProofV3(
        authKey: ByteArray,
        studioId: String,
        nonceD: ByteArray,
        nonceS: ByteArray,
        ephemeralD: ByteArray,
        ephemeralS: ByteArray,
        pairingRequired: Boolean,
        deviceAlias: String,
        mode: Int,
        sessionCounter: Long,
    ): ByteArray = hmac(
        authKey,
        authTranscriptV3(
            "wailo/device-proof/v3",
            studioId,
            nonceD,
            nonceS,
            ephemeralD,
            ephemeralS,
            pairingRequired,
            deviceAlias,
            mode,
            sessionCounter,
        ),
    )

    fun studioProofV3(
        authKey: ByteArray,
        studioId: String,
        nonceD: ByteArray,
        nonceS: ByteArray,
        ephemeralD: ByteArray,
        ephemeralS: ByteArray,
        pairingRequired: Boolean,
        deviceAlias: String,
        mode: Int,
        sessionCounter: Long,
        resultCode: Int,
    ): ByteArray = hmac(
        authKey,
        resultTranscriptV3(
            "wailo/studio-proof/v3",
            studioId,
            nonceD,
            nonceS,
            ephemeralD,
            ephemeralS,
            pairingRequired,
            deviceAlias,
            mode,
            sessionCounter,
            resultCode,
        ),
    )

    fun isValidStudioHelloV3(
        signature: ByteArray,
        publicKey: ByteArray,
        studioId: String,
        nonceD: ByteArray,
        nonceS: ByteArray,
        ephemeralD: ByteArray,
        ephemeralS: ByteArray,
        pairingRequired: Boolean,
    ): Boolean = verifyV3(
        signature,
        publicKey,
        helloTranscriptV3(
            "wailo/studio-hello/v3",
            studioId,
            nonceD,
            nonceS,
            ephemeralD,
            ephemeralS,
            pairingRequired,
        ),
    )

    fun isValidResultSignatureV3(
        signature: ByteArray,
        publicKey: ByteArray,
        studioId: String,
        nonceD: ByteArray,
        nonceS: ByteArray,
        ephemeralD: ByteArray,
        ephemeralS: ByteArray,
        pairingRequired: Boolean,
        deviceAlias: String,
        mode: Int,
        sessionCounter: Long,
        resultCode: Int,
    ): Boolean = verifyV3(
        signature,
        publicKey,
        resultTranscriptV3(
            "wailo/result-signature/v3",
            studioId,
            nonceD,
            nonceS,
            ephemeralD,
            ephemeralS,
            pairingRequired,
            deviceAlias,
            mode,
            sessionCounter,
            resultCode,
        ),
    )

    /** Constant-time, because these compare secrets. */
    fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var difference = 0
        for (index in a.indices) difference = difference or (a[index].toInt() xor b[index].toInt())
        return difference == 0
    }

    // MARK: key encoding

    /**
     * Down to the 65-byte uncompressed point the wire carries. Rebuilt from the affine coordinates
     * rather than sliced out of the DER, so a leading zero in either coordinate cannot shift the result.
     */
    fun encodePublicKey(key: PublicKey): ByteArray {
        val point = (key as ECPublicKey).w
        return byteArrayOf(0x04) + point.affineX.toFixedWidth() + point.affineY.toFixedWidth()
    }

    fun decodePublicKey(encoded: ByteArray): PublicKey {
        require(encoded.size == 1 + 2 * COORDINATE_LENGTH && encoded[0] == 0x04.toByte()) {
            "expected an X9.63 uncompressed P-256 point"
        }
        val x = BigInteger(1, encoded.copyOfRange(1, 1 + COORDINATE_LENGTH))
        val y = BigInteger(1, encoded.copyOfRange(1 + COORDINATE_LENGTH, encoded.size))
        val parameters = curveParameters()
        // The JCA will happily build a key from a point that is not on the curve, and this key then
        // goes into ECDH — the setup for an invalid-curve attack, where an attacker feeds points from
        // weak curves and reads the private scalar out of the answers. CryptoKit rejects these on the
        // other side; match it here rather than rely on the provider.
        require(isOnCurve(x, y, parameters)) { "point is not on P-256" }
        return KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(ECPoint(x, y), parameters))
    }

    private fun isOnCurve(x: BigInteger, y: BigInteger, parameters: ECParameterSpec): Boolean {
        val curve = parameters.curve
        val p = (curve.field as? ECFieldFp)?.p ?: return false
        if (x.signum() < 0 || x >= p || y.signum() < 0 || y >= p) return false
        val left = y.modPow(TWO, p)
        val right = (x.modPow(THREE, p) + curve.a * x + curve.b).mod(p)
        return left == right
    }

    private fun curveParameters(): ECParameterSpec =
        AlgorithmParameters.getInstance("EC")
            .apply { init(ECGenParameterSpec(CURVE)) }
            .getParameterSpec(ECParameterSpec::class.java)

    // MARK: internals

    /** A `0x00`-separated label, so a transcript can never be read two ways. */
    private fun label(text: String, suffix: ByteArray): ByteArray =
        text.toByteArray() + byteArrayOf(0) + suffix + byteArrayOf(0)

    private fun helloTranscriptV3(
        role: String,
        studioId: String,
        nonceD: ByteArray,
        nonceS: ByteArray,
        ephemeralD: ByteArray,
        ephemeralS: ByteArray,
        pairingRequired: Boolean,
    ): ByteArray = label(role, studioId.toByteArray()) +
        nonceD + nonceS + ephemeralD + ephemeralS + byteArrayOf(if (pairingRequired) 1 else 0)

    private fun authTranscriptV3(
        role: String,
        studioId: String,
        nonceD: ByteArray,
        nonceS: ByteArray,
        ephemeralD: ByteArray,
        ephemeralS: ByteArray,
        pairingRequired: Boolean,
        deviceAlias: String,
        mode: Int,
        sessionCounter: Long,
    ): ByteArray = helloTranscriptV3(
        role,
        studioId,
        nonceD,
        nonceS,
        ephemeralD,
        ephemeralS,
        pairingRequired,
    ) + label("wailo/device-alias/v3", deviceAlias.toByteArray()) +
        ByteBuffer.allocate(Int.SIZE_BYTES).putInt(mode).array() +
        ByteBuffer.allocate(Long.SIZE_BYTES).putLong(sessionCounter).array()

    private fun resultTranscriptV3(
        role: String,
        studioId: String,
        nonceD: ByteArray,
        nonceS: ByteArray,
        ephemeralD: ByteArray,
        ephemeralS: ByteArray,
        pairingRequired: Boolean,
        deviceAlias: String,
        mode: Int,
        sessionCounter: Long,
        resultCode: Int,
    ): ByteArray = authTranscriptV3(
        role,
        studioId,
        nonceD,
        nonceS,
        ephemeralD,
        ephemeralS,
        pairingRequired,
        deviceAlias,
        mode,
        sessionCounter,
    ) + ByteBuffer.allocate(Int.SIZE_BYTES).putInt(resultCode).array()

    private fun verifyV3(signature: ByteArray, publicKey: ByteArray, transcript: ByteArray): Boolean = runCatching {
        Signature.getInstance(SIGNATURE_ALGORITHM).apply {
            initVerify(decodePublicKey(publicKey))
            update(transcript)
        }.verify(rawToDerSignature(signature))
    }.getOrDefault(false)

    /** RFC 5869 HKDF-SHA256. One SHA-256 block out, so expand is a single iteration. */
    private fun hkdf(keyMaterial: ByteArray, salt: ByteArray, info: ByteArray): ByteArray =
        hmac(hmac(salt, keyMaterial), info + byteArrayOf(1))

    private fun hmac(key: ByteArray, message: ByteArray): ByteArray =
        Mac.getInstance(HMAC).apply {
            // An all-zero key is legal HMAC input but `SecretKeySpec` rejects an *empty* one, which is
            // what HKDF-extract passes when there is no salt.
            init(SecretKeySpec(if (key.isEmpty()) ByteArray(1) else key, HMAC))
        }.doFinal(message)

    /**
     * RFC 2898 PBKDF2-HMAC-SHA256, hand-composed — and the one place this file has to diverge from the
     * other two implementations, which both call their platform's own.
     *
     * `SecretKeyFactory` only offers `PBKDF2WithHmacSHA256` from Android API 26 against an SDK floor of
     * 24, and vendors are known to strip it even above that. Dropping to `PBKDF2WithHmacSHA1` would be
     * a different function and would silently stop this device pairing with Studio or iOS, and pulling
     * in Bouncy Castle for one KDF is exactly the weight invariant #3 forbids. `Mac` is on every level.
     *
     * The output is 32 bytes, which is one HMAC-SHA256 block, so there is a single `T_1` and no block
     * loop — `WailoCryptoTest` pins this against the JCA's own implementation on the test JVM.
     */
    private fun pbkdf2(password: ByteArray, salt: ByteArray, rounds: Int): ByteArray {
        val mac = Mac.getInstance(HMAC).apply { init(SecretKeySpec(password, HMAC)) }
        var block = mac.doFinal(salt + byteArrayOf(0, 0, 0, 1))
        val result = block.copyOf()
        repeat(rounds - 1) {
            block = mac.doFinal(block)
            for (index in result.indices) result[index] = (result[index].toInt() xor block[index].toInt()).toByte()
        }
        return result
    }

    private fun BigInteger.toFixedWidth(): ByteArray {
        val bytes = toByteArray()
        return when {
            bytes.size == COORDINATE_LENGTH -> bytes
            // BigInteger prepends a sign byte whenever the high bit is set.
            bytes.size > COORDINATE_LENGTH -> bytes.copyOfRange(bytes.size - COORDINATE_LENGTH, bytes.size)
            else -> ByteArray(COORDINATE_LENGTH - bytes.size) + bytes
        }
    }

    /**
     * CryptoKit hands out a bare `r || s`; `SHA256withECDSA` wants `SEQUENCE { INTEGER r, INTEGER s }`.
     * Hand-rolled because this is the entire ASN.1 surface the device needs and the SDK ships inside
     * third-party apps (invariant #3).
     */
    private fun rawToDerSignature(raw: ByteArray): ByteArray {
        require(raw.size == 2 * COORDINATE_LENGTH) { "expected a 64-byte r||s signature" }
        val r = derInteger(BigInteger(1, raw.copyOf(COORDINATE_LENGTH)))
        val s = derInteger(BigInteger(1, raw.copyOfRange(COORDINATE_LENGTH, raw.size)))
        val body = r + s
        return byteArrayOf(0x30, body.size.toByte()) + body
    }

    private fun derInteger(value: BigInteger): ByteArray {
        // BigInteger.toByteArray is already minimal two's complement, which is exactly DER INTEGER.
        val bytes = value.toByteArray()
        return byteArrayOf(0x02, bytes.size.toByte()) + bytes
    }

    private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }
}
